src/dispatch/signal/service.py

```diff
@@ -326,3 +341,110 @@ def supress(
+        signal_instance (SignalInstance): SignalInstance to search for entities in.
+        entity_types (list[EntityType]): List of EntityType objects to search for.
+
+    Returns:
+        list[Entity]: List of Entity objects found.
+
+    Example:
+        >>> signal_instance = SignalInstance(
+        ...     raw={
+        ...         "name": "John Smith",
+        ...         "email": "john.smith@example.com",
+        ...         "phone": "555-555-1212",
+        ...         "address": {
+        ...             "street": "123 Main St",
+        ...             "city": "Anytown",
+        ...             "state": "CA",
+        ...             "zip": "12345"
+        ...         },
+        ...         "notes": "Customer is interested in buying a product."
+        ...     }
+        ... )
+        >>> entity_types = [
+        ...     EntityType(name="Name", field="name", regular_expression=r"\b[A-Z][a-z]+ [A-Z][a-z]+\b"),
+        ...     EntityType(name="Phone", field=None, regular_expression=r"\b\\d{3}[-.]?\\d{3}[-.]?\\d{4}\b"),
+        ... ]
+        >>> entities = find_entities(db_session, signal_instance, entity_types)
+
+    Notes:
+        This function uses depth-first search to traverse the raw data of the signal instance. It searches for
+        the regular expressions specified in the EntityType objects in the values of the dictionary, list, and
+        string objects encountered during the traversal. The search can be limited to a specific key in the
+        dictionary objects by specifying a value for the 'field' attribute of the EntityType object.
+    """
+
+    def search(key, val, entity_type_pairs):
+        # Create a list to hold any entities that are found in this value
+        entities = []
+
+        # If this value has been searched before, return the cached entities
+        if id(val) in cache:
+            return cache[id(val)]
+
+        # If the value is a dictionary, search its key-value pairs recursively
+        if isinstance(val, dict):
+            for subkey, subval in val.items():
+                entities.extend(search(subkey, subval, entity_type_pairs))
+
+        # If the value is a list, search its items recursively
+        elif isinstance(val, list):
+            for item in val:
+                entities.extend(search(None, item, entity_type_pairs))
+
+        # If the value is a string, search it for entity matches
+        elif isinstance(val, str):
+            for entity_type, entity_regex, field in entity_type_pairs:
+                # If a field was specified for this entity type, only search that field
+                if not field or key == field:
+                    # Search the string for matches to the entity type's regular expression
+                    if match := entity_regex.search(val):
+                        # If a match was found, create a new Entity object for it
+                        entity = EntityCreate(
+                            value=match.group(0),
```

Comment:
We might want to add the fact that we don't really support capture groups to the regex tooltip until we figure out how we can let users add them.